File: java/BFS.java
package java;

import java.util.Queue;
import java.util.Scanner;
import java.util.LinkedList;

public class BFS {

    int N; // number of vertices in the graph
    boolean[][] G; // the graph as an adjacency matrix
    // G[i][j] is true if there is an edge from i to j

    boolean[] Acc; // a nodes takes a true value when it is visited or added to the queue
    Queue<Integer> Q;

    BFS(int size) {
        N = size;
        Acc = new boolean[N];
        Q = new LinkedList<Integer>(); // A queue to process nodes
        setupGraph(size);
        System.out.println("---------- The BFS, as follows --------------------");
        System.out.println();
    }

    void setupGraph(int size) {
        G = new boolean[N][N];
        G[0][1] = G[1][0] = true;
        G[0][8] = G[8][0] = true;
        G[0][4] = G[4][0] = true;
        G[1][2] = G[2][1] = true;
        G[1][3] = G[3][1] = true;
        G[2][6] = G[6][2] = true;
        G[3][4] = G[4][3] = true;
        G[3][5] = G[5][3] = true;
        G[6][7] = G[7][6] = true;
        G[8][9] = G[9][8] = true;
        G[9][10] = G[10][9] = true;
        G[10][11] = G[11][10] = true;
    }

    // perform a BFS starting at node start
    void BFS(int start, int g) {
        String addToQu = " ";
        // start with only the start node in the queue and mark it as visited
        Q.offer(start);
        Acc[start] = true;
        System.out.println("The content of the queue as follows:");
        System.out.print("[");
        System.out.print(retCity(Q.peek()));// shows the first element in the queue
        System.out.println("]\n");

        // continue searching the graph while still nodes in the queue
        while (!Q.isEmpty()) {
            int at = Q.poll(); // get the head of the queue
            if (g == at) {
                System.out.println("\nThe goal (" + retCity(g) + ")has been reached\n\n");
                return;
            }
            System.out.println("\n-->At " + retCity(at) + " city now");
            addToQu = retCity(at) + " has not been the goal,children [";

            // add any unseen nodes to the queue to process, then mark them as visited so
            // they don't get re-added

            for (int i = 0; i < N; ++i)
                if (G[at][i] && !Acc[i]) {
                    Q.offer(i);
                    addToQu += retCity(i) + " ";
                    Acc[i] = true;
                }

            addToQu += "]'v been added to the queue";
            System.out.println(addToQu);

            System.out.print("The Queue: [(Front)");
            for (int i : Q)
                System.out.print(".." + retCity(i));
            System.out.print("]\n");

        }

        System.out.printf("Finished with the BFS from start node %d%n, without reaching the goal", start);
        System.exit(0);
    }

    public static void main(String[] args) {
        int cityChoice, goal;
        Scanner inp = new Scanner(System.in);
        System.out.println("\n\nChoose a city to start with(its number): \n");
        for (int i = 0; i < 12; i++)
            System.out.println(retCity(i) + " city[" + i + "]");
        System.out.print("\nInput: ");
        cityChoice = inp.nextInt();
        System.out.println("\nChoose a city to be the goal(its number): ");
        System.out.print("\nInput: ");
        goal = inp.nextInt();

        if (cityChoice < 0 || cityChoice >= 12) {
            System.out.println("Mistake,run the program again");
            System.exit(0);
        }

        BFS schBFS = new BFS(12);
        schBFS.BFS(cityChoice, goal);
    }

    public static String retCity(int i) // the function returns city name, according to its index in V array
    {
        if (i == 0)
            return "Buraydah";
        else if (i == 1)
            return "Unayzah";
        else if (i == 2)
            return "AlZulfi";
        else if (i == 3)
            return "Al-Badai";
        else if (i == 4)
            return "Riyadh-Alkhabra";
        else if (i == 5)
            return "AlRass";
        else if (i == 6)
            return "UmSedrah";
        else if (i == 7)
            return "Shakra";
        else if (i == 8)
            return "Al-Bukayriyah";
        else if (i == 9)
            return "Sheehyah";
        else if (i == 10)
            return "Dhalfa";
        else
            return "Mulida";

    }
}

File: BFS.py
class BFS:
    def __init__(self, n):
        self.N = n
        self.G = [[False for x in range(self.N)] for y in range(self.N)]
        self.Acc = [False] * self.N
        self.Q = []
        self.setupGraph()

    def setupGraph(self):
        self.G[1][0] = self.G[0][1] = True
        self.G[0][5] = self.G[5][0] = True
        self.G[0][6] = self.G[6][0] = True
        self.G[1][2] = self.G[2][1] = True
        self.G[1][3] = self.G[3][1] = True
        self.G[1][6] = self.G[6][1] = True
        self.G[1][9] = self.G[9][1] = True
        self.G[2][3] = self.G[3][2] = True
        self.G[2][4] = self.G[4][2] = True
        self.G[2][6] = self.G[6][2] = True
        self.G[3][9] = self.G[9][3] = True
        self.G[4][6] = self.G[6][4] = True
        self.G[4][7] = self.G[7][4] = True
        self.G[5][6] = self.G[6][5] = True
        self.G[5][7] = self.G[7][5] = True
        self.G[6][7] = self.G[7][6] = True
        self.G[7][8] = self.G[8][7] = True

    def BFS_(self, start, g):
        print("---------- The BFS, as follows --------------------")
        addToQu = " "
        self.Q.append(start)
        self.Acc[start] = True
        print("The content of the queue as follows:")
        print(f"[{self.retCity(self.Q[-1])}]\n")

        while self.Q:
            at = self.Q.pop(0)
            if g == at:
                print(f"\nThe goal (  {self.retCity(g)}  )has been reached\n\n")
                return

            print(f"\n-->At  {self.retCity(at)}  city now")
            addToQu = self.retCity(at) + " has not been the goal,children ["

            for i in range(self.N):
                if self.G[at][i] and not self.Acc[i]:
                    self.Q.append(i)
                    addToQu += self.retCity(i) + " "
                    self.Acc[i] = True

            addToQu += "]'v been added to the queue"
            print(addToQu)

            print("The Queue: [(Front)", end="")
            for i in self.Q:
                print(f"..{self.retCity(i)}", end="")
            print("]")

        print(
            f"Finished with the BFS from start node {start}, without reaching the goal"
        )
        return

    def retCity(self, i):
        if i == 0:
            return "Business and Economics"
        elif i == 1:
            return "Science"
        elif i == 2:
            return "Engineering"
        elif i == 3:
            return "Architecture and Planning"
        elif i == 4:
            return "Sharia and Islamic Studies"
        elif i == 5:
            return "Medicine"
        elif i == 6:
            return "Qassim University Public Library"
        elif i == 7:
            return "Department of English & Translation"
        elif i == 8:
            return "Parking lot 1"
        else:
            return "Parking lot 2"


def main():

    schBFS = BFS(10)
    print("\n\nChoose a city to start with(its number): \n")
    for i in range(10):
        print(f"{schBFS.retCity(i)} city[ {i} ]")
    cityChoice = int(input("\nInput: "))
    print("\nChoose a city to be the goal(its number): ")
    goal = int(input("\nInput: "))

    if cityChoice < 0 or cityChoice >= 10:
        print("Mistake,run the program again")
        return
    schBFS.BFS_(cityChoice, goal)


if __name__ == "__main__":
    main()